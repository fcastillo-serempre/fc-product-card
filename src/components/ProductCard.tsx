import React, { createContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';

import {
  ProductContextProps,
  Product,
  onChangeArgs,
  initialValues,
  ProductCardHandlers,
} from '../interfaces/interfaces';

import { ProductImage } from './ProductImage';
import { ProductTitle } from './ProductTitle';
import { ProductButtons } from './ProductButtons';

import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  // children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  className?: string;
  initialValues?: initialValues;
  onChange?: (args: onChangeArgs) => void;
  product: Product;
  style?: CSSProperties;
  value?: number;
}

export const ProductCard = ({
  children,
  className,
  initialValues,
  onChange,
  product,
  style,
  value,
}: Props) => {
  const {
    counter,
    increaseBy,
    maxCount,
    isMaxCountReached,
    reset,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  });

  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount,
          product,

          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};

ProductCard.Title = ProductTitle;
ProductCard.Image = ProductImage;
ProductCard.Buttons = ProductButtons;
