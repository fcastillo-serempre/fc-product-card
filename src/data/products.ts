import { Product } from '../interfaces/interfaces';

const product1 = {
  id: '1',
  title: 'Coffee Mug - Sin Imagen',
};

const product2 = {
  id: '2',
  img: './coffee-mug2.png',
  title: 'Coffee Mug - Con Imagen',
};

export const products: Product[] = [product1, product2];
