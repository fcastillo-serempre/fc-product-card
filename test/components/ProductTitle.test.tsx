import React from 'react';
import renderer from 'react-test-renderer';
import { ProductCard } from '../../src/components/ProductCard';
import { ProductTitle } from '../../src/components/ProductTitle';

import { products } from '../../src/data/products';

const product = products[0];

describe('iProductTitle', () => {
  test('debe mostrar el componente correctament con el tÃ­tulo personalizado', () => {
    const wrapper = renderer.create(
      <ProductTitle title="Custom product" className="custom-class" />
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe mostrar el componente con el nombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product}>{() => <ProductTitle />}</ProductCard>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
