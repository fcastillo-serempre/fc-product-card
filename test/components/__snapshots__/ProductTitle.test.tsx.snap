// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iProductTitle debe mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - Sin Imagen
  </span>
</div>
`;

exports[`iProductTitle debe mostrar el componente correctament con el t√≠tulo personalizado 1`] = `
<span
  className="productDescription custom-class"
>
  Custom product
</span>
`;
